stan/math/rev/fun/mdivide_left_tri.hpp

```diff
@@ -389,6 +389,70 @@ mdivide_left_tri(const T1 &A, const T2 &b) {
   return res;
 }
 
+template <Eigen::UpLoType TriView, typename T1, typename T2,
+          require_all_matrix_t<T1, T2> * = nullptr,
+          require_any_var_matrix_t<T1, T2> * = nullptr>
+inline auto mdivide_left_tri(const T1 &A, const T2 &B) {
+  using ret_val_type = plain_type_t<decltype(value_of(A) * value_of(B))>;
+  using ret_type = var_value<ret_val_type>;
+
+  if (A.size() == 0) {
+    return ret_type(ret_val_type(0, B.cols()));
+  }
+
+  check_square("mdivide_left_tri", "A", A);
+  check_multiplicable("mdivide_left_tri", "A", A, "B", B);
+
+  if (!is_constant<T1>::value && !is_constant<T2>::value) {
+    auto arena_A = to_var_value(forward_as<promote_scalar_t<var, T1>>(A));
```

Comment:
This is a bit of a mess and is something I could use help on.

The reason this is here is if you replace this with the code that we want:

```
arena_t<promote_scalar_t<var, T1>> arena_A = A;
```

Then lines like the following give errors when `decltype(arena_A)` is a `mat<var>` type:
```
arena_A.val().template triangularView<TriView>().solve(arena_B.val());
```

This is kindof annoying to work around because that `arena_A` is a `mat<var>` type if A is either a `MatrixXd` or a `Matrix<var, -1, -1>`.

Reply:
Like this:

```
lib/eigen_3.3.7/Eigen/src/Core/SolveTriangular.h:102:41: error: no matching member function for call to 'coeffRef'
      ::run(size, othersize, &actualLhs.coeffRef(0,0), actualLhs.outerStride(), &rhs.coeffRef(0,0), rhs.outerStride(), blocking);
                              ~~~~~~~~~~^~~~~~~~
lib/eigen_3.3.7/Eigen/src/Core/SolveTriangular.h:182:18: note: in instantiation of member function
      'Eigen::internal::triangular_solver_selector<Eigen::CwiseUnaryView<Eigen::MatrixBase<Eigen::Map<Eigen::Matrix<stan::math::var_value<double, void>, -1, -1, 0, -1, -1>,
      0, Eigen::Stride<0, 0> > >::val_Op, Eigen::Map<Eigen::Matrix<stan::math::var_value<double, void>, -1, -1, 0, -1, -1>, 0, Eigen::Stride<0, 0> > >,
      Eigen::Map<Eigen::Matrix<double, -1, -1, 0, -1, -1>, 0, Eigen::Stride<0, 0> >, 1, 1, 0, -1>::run' requested here
    Side, Mode>::run(derived().nestedExpression(), otherCopy);
                 ^
lib/eigen_3.3.7/Eigen/src/Core/TriangularMatrix.h:511:14: note: in instantiation of function template specialization
      'Eigen::TriangularViewImpl<Eigen::CwiseUnaryView<Eigen::MatrixBase<Eigen::Map<Eigen::Matrix<stan::math::var_value<double, void>, -1, -1, 0, -1, -1>, 0,
      Eigen::Stride<0, 0> > >::val_Op, Eigen::Map<Eigen::Matrix<stan::math::var_value<double, void>, -1, -1, 0, -1, -1>, 0, Eigen::Stride<0, 0> > >, 1,
      Eigen::Dense>::solveInPlace<1, Eigen::Map<Eigen::Matrix<double, -1, -1, 0, -1, -1>, 0, Eigen::Stride<0, 0> > >' requested here
    { return solveInPlace<OnTheLeft>(other); }
             ^
lib/eigen_3.3.7/Eigen/src/Core/TriangularMatrix.h:541:13: note: in instantiation of function template specialization
      'Eigen::TriangularViewImpl<Eigen::CwiseUnaryView<Eigen::MatrixBase<Eigen::Map<Eigen::Matrix<stan::math::var_value<double, void>, -1, -1, 0, -1, -1>, 0,
      Eigen::Stride<0, 0> > >::val_Op, Eigen::Map<Eigen::Matrix<stan::math::var_value<double, void>, -1, -1, 0, -1, -1>, 0, Eigen::Stride<0, 0> > >, 1,
      Eigen::Dense>::solveInPlace<Eigen::Map<Eigen::Matrix<double, -1, -1, 0, -1, -1>, 0, Eigen::Stride<0, 0> > >' requested here
      this->solveInPlace(dst);
            ^
lib/eigen_3.3.7/Eigen/src/Core/Solve.h:147:15: note: in instantiation of function template specialization
      'Eigen::TriangularViewImpl<Eigen::CwiseUnaryView<Eigen::MatrixBase<Eigen::Map<Eigen::Matrix<stan::math::var_value<double, void>, -1, -1, 0, -1, -1>, 0,
      Eigen::Stride<0, 0> > >::val_Op, Eigen::Map<Eigen::Matrix<stan::math::var_value<double, void>, -1, -1, 0, -1, -1>, 0, Eigen::Stride<0, 0> > >, 1,
      Eigen::Dense>::_solve_impl<Eigen::Map<Eigen::Matrix<double, -1, -1, 0, -1, -1>, 0, Eigen::Stride<0, 0> >, Eigen::Map<Eigen::Matrix<double, -1, -1, 0, -1, -1>, 0,
      Eigen::Stride<0, 0> > >' requested here
    src.dec()._solve_impl(src.rhs(), dst);
              ^
lib/eigen_3.3.7/Eigen/src/Core/AssignEvaluator.h:836:46: note: in instantiation of member function 'Eigen::internal::Assignment<Eigen::Map<Eigen::Matrix<double, -1, -1, 0,
      -1, -1>, 0, Eigen::Stride<0, 0> >, Eigen::Solve<Eigen::TriangularView<Eigen::CwiseUnaryView<Eigen::MatrixBase<Eigen::Map<Eigen::Matrix<stan::math::var_value<double,
      void>, -1, -1, 0, -1, -1>, 0, Eigen::Stride<0, 0> > >::val_Op, Eigen::Map<Eigen::Matrix<stan::math::var_value<double, void>, -1, -1, 0, -1, -1>, 0, Eigen::Stride<0, 0>
      > >, 1>, Eigen::Map<Eigen::Matrix<double, -1, -1, 0, -1, -1>, 0, Eigen::Stride<0, 0> > >, Eigen::internal::assign_op<double, double>, Eigen::internal::Dense2Dense,
      void>::run' requested here
  Assignment<ActualDstTypeCleaned,Src,Func>::run(actualDst, src, func);
                                             ^
lib/eigen_3.3.7/Eigen/src/Core/AssignEvaluator.h:804:3: note: in instantiation of function template specialization
      'Eigen::internal::call_assignment_no_alias<Eigen::Map<Eigen::Matrix<double, -1, -1, 0, -1, -1>, 0, Eigen::Stride<0, 0> >,
      Eigen::Solve<Eigen::TriangularView<Eigen::CwiseUnaryView<Eigen::MatrixBase<Eigen::Map<Eigen::Matrix<stan::math::var_value<double, void>, -1, -1, 0, -1, -1>, 0,
      Eigen::Stride<0, 0> > >::val_Op, Eigen::Map<Eigen::Matrix<stan::math::var_value<double, void>, -1, -1, 0, -1, -1>, 0, Eigen::Stride<0, 0> > >, 1>,
      Eigen::Map<Eigen::Matrix<double, -1, -1, 0, -1, -1>, 0, Eigen::Stride<0, 0> > >, Eigen::internal::assign_op<double, double> >' requested here
  call_assignment_no_alias(dst, src, func);
  ^
lib/eigen_3.3.7/Eigen/src/Core/AssignEvaluator.h:782:3: note: in instantiation of function template specialization
      'Eigen::internal::call_assignment<Eigen::Map<Eigen::Matrix<double, -1, -1, 0, -1, -1>, 0, Eigen::Stride<0, 0> >,
      Eigen::Solve<Eigen::TriangularView<Eigen::CwiseUnaryView<Eigen::MatrixBase<Eigen::Map<Eigen::Matrix<stan::math::var_value<double, void>, -1, -1, 0, -1, -1>, 0,
      Eigen::Stride<0, 0> > >::val_Op, Eigen::Map<Eigen::Matrix<stan::math::var_value<double, void>, -1, -1, 0, -1, -1>, 0, Eigen::Stride<0, 0> > >, 1>,
      Eigen::Map<Eigen::Matrix<double, -1, -1, 0, -1, -1>, 0, Eigen::Stride<0, 0> > >, Eigen::internal::assign_op<double, double> >' requested here
  call_assignment(dst, src, internal::assign_op<typename Dst::Scalar,typename Src::Scalar>());
  ^
lib/eigen_3.3.7/Eigen/src/Core/Assign.h:66:13: note: in instantiation of function template specialization 'Eigen::internal::call_assignment<Eigen::Map<Eigen::Matrix<double,
      -1, -1, 0, -1, -1>, 0, Eigen::Stride<0, 0> >,
      Eigen::Solve<Eigen::TriangularView<Eigen::CwiseUnaryView<Eigen::MatrixBase<Eigen::Map<Eigen::Matrix<stan::math::var_value<double, void>, -1, -1, 0, -1, -1>, 0,
      Eigen::Stride<0, 0> > >::val_Op, Eigen::Map<Eigen::Matrix<stan::math::var_value<double, void>, -1, -1, 0, -1, -1>, 0, Eigen::Stride<0, 0> > >, 1>,
      Eigen::Map<Eigen::Matrix<double, -1, -1, 0, -1, -1>, 0, Eigen::Stride<0, 0> > > >' requested here
  internal::call_assignment(derived(), other.derived());
            ^
lib/eigen_3.3.7/Eigen/src/Core/Map.h:162:5: note: in instantiation of function template specialization 'Eigen::MatrixBase<Eigen::Map<Eigen::Matrix<double, -1, -1, 0, -1,
      -1>, 0, Eigen::Stride<0, 0> >
      >::operator=<Eigen::Solve<Eigen::TriangularView<Eigen::CwiseUnaryView<Eigen::MatrixBase<Eigen::Map<Eigen::Matrix<stan::math::var_value<double, void>, -1, -1, 0, -1,
      -1>, 0, Eigen::Stride<0, 0> > >::val_Op, Eigen::Map<Eigen::Matrix<stan::math::var_value<double, void>, -1, -1, 0, -1, -1>, 0, Eigen::Stride<0, 0> > >, 1>,
      Eigen::Map<Eigen::Matrix<double, -1, -1, 0, -1, -1>, 0, Eigen::Stride<0, 0> > > >' requested here
    EIGEN_INHERIT_ASSIGNMENT_OPERATORS(Map)
    ^
lib/eigen_3.3.7/Eigen/src/Core/util/Macros.h:842:53: note: expanded from macro 'EIGEN_INHERIT_ASSIGNMENT_OPERATORS'
#define EIGEN_INHERIT_ASSIGNMENT_OPERATORS(Derived) EIGEN_INHERIT_ASSIGNMENT_EQUAL_OPERATOR(Derived)
                                                    ^
lib/eigen_3.3.7/Eigen/src/Core/util/Macros.h:826:108: note: expanded from macro 'EIGEN_INHERIT_ASSIGNMENT_EQUAL_OPERATOR'
    EIGEN_DEVICE_FUNC EIGEN_STRONG_INLINE Derived& operator=(const DenseBase<OtherDerived>& other) { Base::operator=(other.derived()); return *this; }
                                                                                                           ^
./stan/math/rev/core/arena_matrix.hpp:109:11: note: in instantiation of function template specialization 'Eigen::Map<Eigen::Matrix<double, -1, -1, 0, -1, -1>, 0,
      Eigen::Stride<0, 0> >::operator=<Eigen::Solve<Eigen::TriangularView<Eigen::CwiseUnaryView<Eigen::MatrixBase<Eigen::Map<Eigen::Matrix<stan::math::var_value<double,
      void>, -1, -1, 0, -1, -1>, 0, Eigen::Stride<0, 0> > >::val_Op, Eigen::Map<Eigen::Matrix<stan::math::var_value<double, void>, -1, -1, 0, -1, -1>, 0, Eigen::Stride<0, 0>
      > >, 1>, Eigen::Map<Eigen::Matrix<double, -1, -1, 0, -1, -1>, 0, Eigen::Stride<0, 0> > > >' requested here
    Base::operator=(a);
          ^
./stan/math/rev/core/arena_matrix.hpp:65:11: note: in instantiation of function template specialization 'stan::math::arena_matrix<Eigen::Matrix<double, -1, -1, 0, -1, -1>
      >::operator=<Eigen::Solve<Eigen::TriangularView<Eigen::CwiseUnaryView<Eigen::MatrixBase<Eigen::Map<Eigen::Matrix<stan::math::var_value<double, void>, -1, -1, 0, -1,
      -1>, 0, Eigen::Stride<0, 0> > >::val_Op, Eigen::Map<Eigen::Matrix<stan::math::var_value<double, void>, -1, -1, 0, -1, -1>, 0, Eigen::Stride<0, 0> > >, 1>,
      Eigen::Map<Eigen::Matrix<double, -1, -1, 0, -1, -1>, 0, Eigen::Stride<0, 0> > > >' requested here
    *this = other;
          ^
./stan/math/rev/core/vari.hpp:624:42: note: in instantiation of function template specialization 'stan::math::arena_matrix<Eigen::Matrix<double, -1, -1, 0, -1, -1>
      >::arena_matrix<Eigen::Solve<Eigen::TriangularView<Eigen::CwiseUnaryView<Eigen::MatrixBase<Eigen::Map<Eigen::Matrix<stan::math::var_value<double, void>, -1, -1, 0, -1,
      -1>, 0, Eigen::Stride<0, 0> > >::val_Op, Eigen::Map<Eigen::Matrix<stan::math::var_value<double, void>, -1, -1, 0, -1, -1>, 0, Eigen::Stride<0, 0> > >, 1>,
      Eigen::Map<Eigen::Matrix<double, -1, -1, 0, -1, -1>, 0, Eigen::Stride<0, 0> > >, nullptr>' requested here
  vari_value(const S& x, bool stacked) : val_(x), adj_(x.rows(), x.cols()) {
                                         ^
./stan/math/rev/core/var.hpp:363:30: note: in instantiation of function template specialization 'stan::math::vari_value<Eigen::Matrix<double, -1, -1, 0, -1, -1>,
      void>::vari_value<Eigen::Solve<Eigen::TriangularView<Eigen::CwiseUnaryView<Eigen::MatrixBase<Eigen::Map<Eigen::Matrix<stan::math::var_value<double, void>, -1, -1, 0,
      -1, -1>, 0, Eigen::Stride<0, 0> > >::val_Op, Eigen::Map<Eigen::Matrix<stan::math::var_value<double, void>, -1, -1, 0, -1, -1>, 0, Eigen::Stride<0, 0> > >, 1>,
      Eigen::Map<Eigen::Matrix<double, -1, -1, 0, -1, -1>, 0, Eigen::Stride<0, 0> > >, nullptr>' requested here
  var_value(S&& x) : vi_(new vari_type(std::forward<S>(x), false)) {}  // NOLINT
                             ^
./stan/math/rev/fun/mdivide_left_tri.hpp:428:9: note: in instantiation of function template specialization 'stan::math::var_value<Eigen::Matrix<double, -1, -1, 0, -1, -1>,
      void>::var_value<const Eigen::Solve<Eigen::TriangularView<Eigen::CwiseUnaryView<Eigen::MatrixBase<Eigen::Map<Eigen::Matrix<stan::math::var_value<double, void>, -1, -1,
      0, -1, -1>, 0, Eigen::Stride<0, 0> > >::val_Op, Eigen::Map<Eigen::Matrix<stan::math::var_value<double, void>, -1, -1, 0, -1, -1>, 0, Eigen::Stride<0, 0> > >, 1>,
      Eigen::Map<Eigen::Matrix<double, -1, -1, 0, -1, -1>, 0, Eigen::Stride<0, 0> > >, nullptr>' requested here
      = arena_A.val().template triangularView<TriView>().solve(value_of(B));
```

Reply:
This is due to the same thing as multiplying the non-const ```val()```s. Can you try with ````val_op``` and see if it works?

Reply:
Ah okay, will try that

Reply:
@andrjohns had a go with `val_op` and `adj_op` and I'm still getting errors.

So with this code:
```cpp
    using ret_val_type = plain_type_t<decltype(value_of(A) * value_of(B))>;
    using ret_type = var_value<ret_val_type>;

    arena_t<promote_scalar_t<var, T1>> arena_A = A;
    arena_t<promote_scalar_t<double, T2>> arena_B = value_of(B);

    ret_val_type res_val = arena_A.val_op().template triangularView<TriView>().solve(arena_B);
    arena_t<ret_type> res = res_val;
```

I'm getting this error:
```
lib/eigen_3.3.7/Eigen/src/Core/SolveTriangular.h:73:41: error: no member named 'data' in 'Eigen::CwiseUnaryOp<Eigen::MatrixBase<Eigen::Map<Eigen::Matrix<stan::math::var_value<double, void>, -1, -1, 0, -1, -1>, 0, Eigen::Stride<0, 0> > >::val_Op,
      Eigen::Map<Eigen::Matrix<stan::math::var_value<double, void>, -1, -1, 0, -1, -1>, 0, Eigen::Stride<0, 0> > >'
      ::run(actualLhs.cols(), actualLhs.data(), actualLhs.outerStride(), actualRhs);
                              ~~~~~~~~~ ^
lib/eigen_3.3.7/Eigen/src/Core/SolveTriangular.h:182:18: note: in instantiation of member function 'Eigen::internal::triangular_solver_selector<Eigen::CwiseUnaryOp<Eigen::MatrixBase<Eigen::Map<Eigen::Matrix<stan::math::var_value<double, void>, -1, -1, 0, -1, -1>, 0,
      Eigen::Stride<0, 0> > >::val_Op, Eigen::Map<Eigen::Matrix<stan::math::var_value<double, void>, -1, -1, 0, -1, -1>, 0, Eigen::Stride<0, 0> > >, Eigen::Matrix<double, -1, 1, 0, -1, 1>, 1, 1, 0, 1>::run' requested here
    Side, Mode>::run(derived().nestedExpression(), otherCopy);
                 ^
lib/eigen_3.3.7/Eigen/src/Core/TriangularMatrix.h:511:14: note: in instantiation of function template specialization 'Eigen::TriangularViewImpl<Eigen::CwiseUnaryOp<Eigen::MatrixBase<Eigen::Map<Eigen::Matrix<stan::math::var_value<double, void>, -1, -1, 0, -1, -1>, 0,
      Eigen::Stride<0, 0> > >::val_Op, Eigen::Map<Eigen::Matrix<stan::math::var_value<double, void>, -1, -1, 0, -1, -1>, 0, Eigen::Stride<0, 0> > >, 1, Eigen::Dense>::solveInPlace<1, Eigen::Matrix<double, -1, 1, 0, -1, 1> >' requested here
    { return solveInPlace<OnTheLeft>(other); }
             ^
lib/eigen_3.3.7/Eigen/src/Core/TriangularMatrix.h:541:13: note: in instantiation of function template specialization 'Eigen::TriangularViewImpl<Eigen::CwiseUnaryOp<Eigen::MatrixBase<Eigen::Map<Eigen::Matrix<stan::math::var_value<double, void>, -1, -1, 0, -1, -1>, 0,
      Eigen::Stride<0, 0> > >::val_Op, Eigen::Map<Eigen::Matrix<stan::math::var_value<double, void>, -1, -1, 0, -1, -1>, 0, Eigen::Stride<0, 0> > >, 1, Eigen::Dense>::solveInPlace<Eigen::Matrix<double, -1, 1, 0, -1, 1> >' requested here
      this->solveInPlace(dst);
            ^
lib/eigen_3.3.7/Eigen/src/Core/Solve.h:147:15: note: in instantiation of function template specialization 'Eigen::TriangularViewImpl<Eigen::CwiseUnaryOp<Eigen::MatrixBase<Eigen::Map<Eigen::Matrix<stan::math::var_value<double, void>, -1, -1, 0, -1, -1>, 0,
      Eigen::Stride<0, 0> > >::val_Op, Eigen::Map<Eigen::Matrix<stan::math::var_value<double, void>, -1, -1, 0, -1, -1>, 0, Eigen::Stride<0, 0> > >, 1, Eigen::Dense>::_solve_impl<Eigen::Map<Eigen::Matrix<double, -1, 1, 0, -1, 1>, 0, Eigen::Stride<0, 0> >,
      Eigen::Matrix<double, -1, 1, 0, -1, 1> >' requested here
    src.dec()._solve_impl(src.rhs(), dst);
              ^
lib/eigen_3.3.7/Eigen/src/Core/AssignEvaluator.h:836:46: note: in instantiation of member function 'Eigen::internal::Assignment<Eigen::Matrix<double, -1, 1, 0, -1, 1>,
      Eigen::Solve<Eigen::TriangularView<Eigen::CwiseUnaryOp<Eigen::MatrixBase<Eigen::Map<Eigen::Matrix<stan::math::var_value<double, void>, -1, -1, 0, -1, -1>, 0, Eigen::Stride<0, 0> > >::val_Op, Eigen::Map<Eigen::Matrix<stan::math::var_value<double, void>, -1, -1, 0,
      -1, -1>, 0, Eigen::Stride<0, 0> > >, 1>, Eigen::Map<Eigen::Matrix<double, -1, 1, 0, -1, 1>, 0, Eigen::Stride<0, 0> > >, Eigen::internal::assign_op<double, double>, Eigen::internal::Dense2Dense, void>::run' requested here
  Assignment<ActualDstTypeCleaned,Src,Func>::run(actualDst, src, func);
                                             ^
lib/eigen_3.3.7/Eigen/src/Core/PlainObjectBase.h:732:17: note: in instantiation of function template specialization 'Eigen::internal::call_assignment_no_alias<Eigen::Matrix<double, -1, 1, 0, -1, 1>,
      Eigen::Solve<Eigen::TriangularView<Eigen::CwiseUnaryOp<Eigen::MatrixBase<Eigen::Map<Eigen::Matrix<stan::math::var_value<double, void>, -1, -1, 0, -1, -1>, 0, Eigen::Stride<0, 0> > >::val_Op, Eigen::Map<Eigen::Matrix<stan::math::var_value<double, void>, -1, -1, 0,
      -1, -1>, 0, Eigen::Stride<0, 0> > >, 1>, Eigen::Map<Eigen::Matrix<double, -1, 1, 0, -1, 1>, 0, Eigen::Stride<0, 0> > >, Eigen::internal::assign_op<double, double> >' requested here
      internal::call_assignment_no_alias(this->derived(), other.derived(), internal::assign_op<Scalar,typename OtherDerived::Scalar>());
                ^
lib/eigen_3.3.7/Eigen/src/Core/PlainObjectBase.h:537:7: note: in instantiation of function template specialization 'Eigen::PlainObjectBase<Eigen::Matrix<double, -1, 1, 0, -1, 1>
      >::_set_noalias<Eigen::Solve<Eigen::TriangularView<Eigen::CwiseUnaryOp<Eigen::MatrixBase<Eigen::Map<Eigen::Matrix<stan::math::var_value<double, void>, -1, -1, 0, -1, -1>, 0, Eigen::Stride<0, 0> > >::val_Op, Eigen::Map<Eigen::Matrix<stan::math::var_value<double,
      void>, -1, -1, 0, -1, -1>, 0, Eigen::Stride<0, 0> > >, 1>, Eigen::Map<Eigen::Matrix<double, -1, 1, 0, -1, 1>, 0, Eigen::Stride<0, 0> > > >' requested here
      _set_noalias(other);
      ^
lib/eigen_3.3.7/Eigen/src/Core/Matrix.h:377:9: note: in instantiation of function template specialization 'Eigen::PlainObjectBase<Eigen::Matrix<double, -1, 1, 0, -1, 1>
      >::PlainObjectBase<Eigen::Solve<Eigen::TriangularView<Eigen::CwiseUnaryOp<Eigen::MatrixBase<Eigen::Map<Eigen::Matrix<stan::math::var_value<double, void>, -1, -1, 0, -1, -1>, 0, Eigen::Stride<0, 0> > >::val_Op, Eigen::Map<Eigen::Matrix<stan::math::var_value<double,
      void>, -1, -1, 0, -1, -1>, 0, Eigen::Stride<0, 0> > >, 1>, Eigen::Map<Eigen::Matrix<double, -1, 1, 0, -1, 1>, 0, Eigen::Stride<0, 0> > > >' requested here
      : Base(other.derived())
        ^
./stan/math/rev/fun/mdivide_left_tri.hpp:429:28: note: in instantiation of function template specialization 'Eigen::Matrix<double, -1, 1, 0, -1,
      1>::Matrix<Eigen::Solve<Eigen::TriangularView<Eigen::CwiseUnaryOp<Eigen::MatrixBase<Eigen::Map<Eigen::Matrix<stan::math::var_value<double, void>, -1, -1, 0, -1, -1>, 0, Eigen::Stride<0, 0> > >::val_Op, Eigen::Map<Eigen::Matrix<stan::math::var_value<double, void>,
      -1, -1, 0, -1, -1>, 0, Eigen::Stride<0, 0> > >, 1>, Eigen::Map<Eigen::Matrix<double, -1, 1, 0, -1, 1>, 0, Eigen::Stride<0, 0> > > >' requested here
    ret_val_type res_val = arena_A.val_op().template triangularView<TriView>().solve(arena_B);
                           ^
./test/unit/math/test_ad.hpp:1912:3: note: in instantiation of function template specialization 'stan::test::expect_ad_matvar_vv<double, stan::math::var_value<double, void>, (lambda at test/unit/math/mix/fun/mdivide_left_tri_test.cpp:4:12), Eigen::Matrix<double, -1, -1,
      0, -1, -1>, Eigen::Matrix<double, -1, 1, 0, -1, 1> >' requested here
  expect_ad_matvar_vv<double, var>(tols, f, x, y);
  ^
./test/unit/math/test_ad.hpp:1935:3: note: in instantiation of function template specialization 'stan::test::expect_ad_matvar<(lambda at test/unit/math/mix/fun/mdivide_left_tri_test.cpp:4:12), Eigen::Matrix<double, -1, -1, 0, -1, -1>, Eigen::Matrix<double, -1, 1, 0, -1,
      1> >' requested here
  expect_ad_matvar(tols, f, x, y);
  ^
test/unit/math/mix/fun/mdivide_left_tri_test.cpp:16:15: note: in instantiation of function template specialization 'stan::test::expect_ad_matvar<(lambda at test/unit/math/mix/fun/mdivide_left_tri_test.cpp:4:12), Eigen::Matrix<double, -1, -1, 0, -1, -1>,
      Eigen::Matrix<double, -1, 1, 0, -1, 1> >' requested here
  stan::test::expect_ad_matvar(f, m00, v0);
```

If I comment out the solve the code runs (leaving the `triangularView` in there). Any ideas? Or do I need to pull out the values into a temporary before I do `.template triangularView<TriView>().solve(arena_B)`?

Reply:
I started looking at this. I'll check it out more on sunday. Ben answer could be ^ but there should be a nice way to do this with our current stuff.

Reply:
Rest of the code looks good tho

Reply:
Found a way to work around this. I just copy out the vals into their own arena variable variable and use that for the vals.

Like:
```
auto arena_A_val = to_arena(arena_A.val());
```

This doesn't cause any extra work for varmats because the vals are an arena_t already.